refactor(forms): make error messages easier to tree shake

crisbeto opened the pull request:
Currently the error message functions are defined as static methods on a class which means that as soon as one of them is used somewhere, all of them have to be retained. This isn't a problem at the moment, because all of them are behind `ngDevMode` checks, but it's error prone and it's easy to fix.

These changes move them out into functions so that they can be imported individually. It also has the advantage of allowing Webpack to minify the function names.

AndrewKushnir reviewed (commented):
Hi @crisbeto,

Thanks for creating this PR.

As you mentioned, the errors are tree-shaken away in prod bundles due to the `ngDevMode` flag and we have golden files (that track which symbols are being retained) to prevent regressions. I think there is a value in this refactoring still, but I'd like to propose a couple things:

- modify the code to move `throw` and `console.log` calls from the helper functions to the higher-level code, so it's clear that we terminate the execution in case there is an error. The helper functions would just create and prepare an instance of an `Error`, i.e. the code may look like this (vs having no `throw` there atm):
```
    if (!this.form && (typeof ngDevMode === 'undefined' || ngDevMode)) {
      throw missingFormException();
    }
```

- another refactoring that would be great to do is to employ `RuntimeError` class that we have in the core package. That way we would bring the error-processing subsystem in Forms to the state consistent with the core (and later on we can expand it to be used in other packages too (for ex. in Router)).

I can help with the above-mentioned changes as well if needed, please let me know.

Thank you.

crisbeto commented:
I've pushed the change to move the `throw`/`console.warn` statements out of the functions, but I didn't do the `RuntimeError` refactoring. The problem is that `RuntimeError` requires an error code and the error code enum is defined in `core`. It might be weird if `core` knows about errors in `forms`.

crisbeto committed:
refactor(forms): make error messages easier to tree shake

Currently the error message functions are defined as static methods on a class which means that as soon as one of them is used somewhere, all of them have to be retained. This isn't a problem at the moment, because all of them are behind `ngDevMode` checks, but it's error prone and it's easy to fix.

These changes move them out into functions so that they can be imported individually. It also has the advantage of allowing Webpack to minify the function names.

AndrewKushnir reviewed (approved):
LGTM, thanks for the refactoring @crisbeto 👍
